### PR TITLE
[OTX][XAI] OTX XAI integration

### PR DESCRIPTION
**Pull request by dongkwan-kim01**
**Summary**
OTX XAI Integration, including explanation mode support for OV/Torch, Classification/Detection and Inferrer patch
This PR needs merge along with MPA xai detection support PR # 85([link](https://github.com/openvinotoolkit/model_preparation_algorithm/pull/85))
Supported Algorithms: `ClassWiseSaliencyMap`(channel-wise mean of the feature map, ReciproCAM for cls case), `EigenCam`, `ActivationMap`
Supported Models:
- All classification models
- All detection models, but two-stage detectors currently not support `ClassWiseSaliencyMap`

**This PR includes**
- Integration of explain mode in the `otx`
- CLI sanity check: `tests/integration/cli`
- API sanity and saliency map validity check: `tests/integration/api/xai`

**Test Codes**
- Run-code
```
otx explain {ROOT_DIR}/otx/algorithms/detection/configs/detection/mobilenetv2_ssd/template.yaml \
--explain-data-roots "{ROOT_DIR}/data/vehicle_detection/train/" \
--save-explanation-to det_xai_outputs \
--load-weights {ROOT_DIR}/results/det_model/weights.pth \
```
- Test-code
```
pytest -rs {ROOT_DIR}/tests/integration/api/xai/
```

### CODE REVIEW

File: requirements/classification.txt [hunk 1]
```diff
@@ -1,3 +1,3 @@
 mmcv-full==1.3.14
 torchreid@git+https://github.com/openvinotoolkit/deep-object-reid@otx
-mpa@git+https://github.com/openvinotoolkit/model_preparation_algorithm@otx
\ No newline at end of file
+mpa@git+https://github.com/openvinotoolkit/model_preparation_algorithm@dk/otx_xai_baseline
```
Comment (wonjuleee) [inline requirements/classification.txt]: Let's merge MPA code change first and leave this by @otx

File: otx/algorithms/classification/adapters/openvino/model_wrappers/openvino_models.py [hunk 2]
```diff
@@ -106,27 +106,21 @@ def postprocess(self, outputs: Dict[str, np.ndarray], meta: Dict[str, Any]):  #
...
-        """Postprocess for aux outputs."""
-        actmap = get_actmap(outputs["saliency_map"][0], (metadata["original_shape"][1], metadata["original_shape"][0]))
+        """Post-process for auxiliary outputs."""
+        saliency_map = outputs["saliency_map"][0]
         repr_vector = outputs["feature_vector"].reshape(-1)
         logits = outputs[self.out_layer_name].squeeze()
         if self.multilabel:
             probs = sigmoid_numpy(logits)
+        elif self.hierarchical:
+            probs = activate_multihead_output(logits, self.multihead_class_info)
```
Comment (dongkwan-kim01) [inline otx/algorithms/classification/adapters/openvino/model_wrappers/openvino_models.py:119]: Followed up by #1386 for active score evaluation

File: otx/algorithms/classification/adapters/openvino/model_wrappers/openvino_models.py [hunk 3]
```diff
@@ -143,6 +137,20 @@ def softmax_numpy(x: np.ndarray):
...
+    """Activate multihead output."""
+    for i in range(multihead_class_info["num_multiclass_heads"]):
+        logits_begin, logits_end = multihead_class_info["head_idx_to_logits_range"][i]
+        logits[logits_begin:logits_end] = softmax_numpy(logits[logits_begin:logits_end])
+
+    if multihead_class_info["num_multilabel_classes"]:
+        logits_begin, logits_end = multihead_class_info["num_single_label_classes"], -1
+        logits[logits_begin:logits_end] = softmax_numpy(logits[logits_begin:logits_end])
+
+    return logits
```
Comment (dongkwan-kim01) [inline otx/algorithms/classification/adapters/openvino/model_wrappers/openvino_models.py:151]: Same as above, #1386

File: otx/cli/utils/tests.py [hunk 4]
```diff
@@ -494,32 +493,59 @@ def otx_explain_testing(template, root, otx_dir, args):
...
+        train_type = "default"
+
+    save_dir = f"explain_{template.model_template_id}/{test_algorithm}/{train_type}/"
+    output_dir = os.path.join(template_work_dir, save_dir)
+    command_line = [
+        "otx",
+        "explain",
+        template.model_template_path,
+        "--load-weights",
+        f"{template_work_dir}/exported_nncf_{template.model_template_id}/openvino.xml",
```
Comment (goodsong81) [inline otx/cli/utils/tests.py]: Why changed to exported NNCF IR?

File: tests/integration/api/xai/test_api_xai_validity.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+from otx.cli.registry import Registry
+from tests.test_suite.e2e_test_system import e2e_pytest_unit
+
+torch.manual_seed(0)
+
+templates_cls = Registry("external/model-preparation-algorithm").filter(task_type="CLASSIFICATION").templates
+templates_cls_ids = [template.model_template_id for template in templates_cls]
+
+templates_det = Registry("external/model-preparation-algorithm").filter(task_type="DETECTION").templates
+templates_det_ids = [template.model_template_id for template in templates_det]
```
Comment (goodsong81) [inline tests/integration/api/xai/test_api_xai_validity.py]: Don't use "external/" directory in feature/otx branch.

**Review by goodsong81 — CHANGES_REQUESTED** [review]
Thank you for the effort.
There are a few things to fix. Could you please revise? (MPA PR, too)
Then, let's see the test result after that.

File: otx/algorithms/classification/tasks/inference.py [hunk 6]
```diff
@@ -352,7 +348,6 @@ def _init_recipe(self):
         # FIXME[Soobee] : if train type is not in cfg, it raises an error in default INCREMENTAL mode.
         # During semi-implementation, this line should be fixed to -> self._recipe_cfg.train_type = train_type
         self._recipe_cfg.train_type = train_type.name
-        patch_data_pipeline(self._recipe_cfg, self.data_pipeline_path)
```
Comment (goodsong81) [inline otx/algorithms/classification/tasks/inference.py]: Could you let me know why you removed this line?
Is this relevant to this PR? OR you just applied some changes from other PRs?

File: tests/integration/api/xai/test_api_xai_validity.py (hunk 5 above)
Comment (dongkwan-kim01) [reply tests/integration/api/xai/test_api_xai_validity.py]: Thank you for the fix!

File: otx/cli/utils/tests.py (hunk 4 above)
Comment (dongkwan-kim01) [reply otx/cli/utils/tests.py]: Yes, nncf is not right. Sorry for the confusion.

File: otx/algorithms/classification/tasks/inference.py [hunk 7]
```diff
@@ -260,31 +259,28 @@ def _add_predictions_to_dataset(self, prediction_results, dataset, update_progre
...
-                    annotation_scene=dataset_item.annotation_scene,
-                    numpy=actmap,
-                    roi=dataset_item.roi,
+                if saliency_map.ndim == 4 and saliency_map.shape[0] == 1:
+                    saliency_map = saliency_map[0]
+                add_saliency_maps_to_dataset_item(
+                    dataset_item=dataset_item,
+                    saliency_map=saliency_map,
+                    model=self._task_environment.model,
+                    labels=self._labels,
```
Comment (goodsong81) [inline otx/algorithms/classification/tasks/inference.py]: Can we move the if block to add_saliency_maps_to_dataset_item()?
Is the condition is common or task specific?

File: otx/algorithms/classification/tasks/openvino.py [hunk 8]
```diff
@@ -221,25 +225,81 @@ def infer(
...
+                                type="saliency_map",
+                                annotation_scene=dataset_item.annotation_scene,
+                                numpy=actmap,
+                                roi=dataset_item.roi,
+                            )
+                            dataset_item.append_metadata_item(saliency_media, model=self.model)
+                else:
+                    raise RuntimeError(
+                        f"Single saliency map has to be 2 or 3-dimensional, " f"but got {saliency_map.ndim} dims"
+                    )
```
Comment (goodsong81) [inline otx/algorithms/classification/tasks/openvino.py]: Can we replace this block by add_saliency_maps_to_dataset_item(), too?

**Review by goodsong81 — COMMENTED** [review]
A few questions.

File: otx/algorithms/classification/tasks/inference.py (hunk 7 above)
Comment (dongkwan-kim01) [reply otx/algorithms/classification/tasks/inference.py]: There were some uncertainty with the input shape, I'll remove this line

File: otx/algorithms/classification/tasks/inference.py (hunk 6 above)
Comment (dongkwan-kim01) [reply otx/algorithms/classification/tasks/inference.py]: This change is because of error, so I followed up develop PR. Sorry for missed the PR number, I'll update when I check it

File: otx/algorithms/classification/tasks/openvino.py (hunk 8 above)
Comment (dongkwan-kim01) [reply otx/algorithms/classification/tasks/openvino.py]: Yes I'll integrate it too

File: tests/integration/api/detection/test_api_detection.py [hunk 9]
```diff
@@ -130,6 +122,20 @@ def setup_configurable_parameters(template_dir, num_iters=10):
         hyper_parameters.postprocessing.confidence_threshold = 0.1
         return hyper_parameters, model_template
 
+
+class TestMPADetAPI(MPADetAPIBase):
```
Comment (negvet) [inline tests/integration/api/detection/test_api_detection.py]: It was renamed from `TestMPADetAPI` to `TestDetectionTaskAPI` for some reason. Let's revert to the new name? 
The base class can be called `DetectionTaskAPIBase`.

**Comment by dongkwan-kim01** [comment]
image
image

Sanity, Validity Test code passed

**Review by negvet — CHANGES_REQUESTED** [review]

File: tests/integration/api/detection/test_api_detection.py (hunk 9 above)
Comment (dongkwan-kim01) [reply tests/integration/api/detection/test_api_detection.py]: Yes, thank you. Same as classification?

Comment (negvet) [reply tests/integration/api/detection/test_api_detection.py]: let's align cls and det for consistency?

File: otx/cli/tools/explain.py [hunk 10]
```diff
@@ -104,7 +107,9 @@ def main():
     hyper_parameters = create(hyper_parameters)
 
     # Get classes for Task, ConfigurableParameters and Dataset.
-    if args.load_weights.endswith(".pth"):
+    if any(args.load_weights.endswith(x) for x in (".bin", ".xml", ".zip")):
+        task_class = get_impl_class(template.entrypoints.openvino)
+    elif args.load_weights.endswith(".pth"):
         if is_checkpoint_nncf(args.load_weights):
```
Comment (negvet) [inline otx/cli/tools/explain.py:113]: BTW, I wonder if we should expect any problems when we have NNCF checkpoint.. Could you please take a look, if have time?
May be do some local explain run from nncf checkpoint..

Comment (dongkwan-kim01) [reply otx/cli/tools/explain.py:113]: Yes, I'll check it.

File: otx/api/utils/dataset_utils.py [hunk 11]
```diff
@@ -183,3 +189,63 @@ def contains_anomalous_images(dataset: DatasetEntity) -> bool:
...
+                )
+            elif task == "det":
+                for bbox in predicted_scene.annotations:
+                    scored_label = bbox.get_labels()[0]
+                    predicted_labels.add(scored_label.label)
+
+        num_saliency_maps = saliency_map.shape[0]
+        if num_saliency_maps == len(labels) + 1:
+            # Include the background as the last category
+            labels.append(LabelEntity("background", Domain.DETECTION))
```
Comment (negvet) [inline otx/api/utils/dataset_utils.py]: This is a common method for cls and det tasks, right?. Can we make some check here (it is better to add this background label only for det task)?

File: otx/algorithms/detection/tasks/openvino.py [hunk 12]
```diff
@@ -340,18 +341,47 @@ def infer(
...
+        logger.info("Start OpenVINO explain")
+
+        update_progress_callback = default_progress_callback
+        if explain_parameters is not None:
+            update_progress_callback = explain_parameters.update_progress  # type: ignore
+        dataset_size = len(dataset)
+        for i, dataset_item in enumerate(dataset, 1):
+            predicted_scene, features = self.inferencer.predict(dataset_item.numpy)
+            dataset_item.append_annotations(predicted_scene.annotations)
+            update_progress_callback(int(i / dataset_size * 100), None)
```
Comment (negvet) [inline otx/algorithms/detection/tasks/openvino.py:371]: May be it is fair to move this progress callback to the bottom of the fool loop body? (like in cls)

**Comment by negvet** [comment]
Please check the tiling detection case. 
E.g. take a look into this fix https://github.com/openvinotoolkit/training_extensions/pull/1408/commits/30b1529ec7030103f2f487cbe09521a7798f6370. Does it applies here?

[not relevant comment - tilling is not supported in OTX yet]

File: otx/algorithms/detection/tasks/openvino.py (hunk 12 above)
Comment (dongkwan-kim01) [reply otx/algorithms/detection/tasks/openvino.py:371]: Right, fixed

**Comment by negvet** [comment]
Except for the above minor comments, the rest LGTM

**Comment by dongkwan-kim01** [comment]
### Comparison between OTE and OTX side inference saliency map

---

image
image

The classification map is almost same(not exactly pixel-wise same though), but detection output is little bit different.

**Review by supersoob — COMMENTED** [review]
Is external/submodule update necessary? If not, please don't include them. Especially ./submodule might came from the mpa local install. Make sure that they are not included.

File: otx/algorithms/classification/tasks/inference.py (hunk 6 above)
Comment (supersoob) [reply otx/algorithms/classification/tasks/inference.py]: Could you go over this?

Comment (dongkwan-kim01) [reply otx/algorithms/classification/tasks/inference.py]: I missed out while following #1431, sorry for the confusion.

**Comment by dongkwan-kim01** [comment]
Classification CLI e2e passed(explain cli sanity check), 3 fail is not related to this PR
image
image

Detection CLI e2e passed(simplified train -> export -> explain(torch), explain(ov) sanity check)
image

xai API sanity check, validity(saliency map validity) passed
image
image

@goodsong81 @negvet @supersoob @wonjuleee @harimkang @sungmanc All comments applied. Could you re-review this PR?

**Review by harimkang — APPROVED** [review]
LGTM
There is an issue in a part unrelated to the PR, so the test cannot be confirmed through the current github action CI. Approved as no problem in local testing.

File: otx/cli/tools/explain.py [hunk 13]
```diff
@@ -76,12 +76,15 @@ def parse_args():
     )
     parser.add_argument(
         "--explain-algorithm",
-        default="EigenCAM",
-        help=f"Explain algorithm name, currently support {SUPPORTED_EXPLAIN_ALGORITHMS}",
+        default="ClassWiseSaliencyMap",
```
Comment (goodsong81) [inline otx/cli/tools/explain.py:79]: Only camel case here?

File: otx/cli/tools/explain.py [hunk 14]
```diff
@@ -76,12 +76,15 @@ def parse_args():
...
+        help=(
+            f"Explain algorithm name, currently support {SUPPORTED_EXPLAIN_ALGORITHMS}.",
```
Comment (goodsong81) [inline otx/cli/tools/explain.py:81]: If you want to show these options to users, better strings are needed.
Maybe, ACTIVATION, EIGENCAM, CLASSWISE_SALIENCY?

**Comment by negvet** [comment]
> ### Comparison between OTE and OTX side inference saliency map
> The classification map is almost same(not exactly pixel-wise same though), but detection output is little bit different.

Thank you for the comparison! This is useful sanity check.

Although the saliency maps should be exactly the same for the same model and data, IMO difference in the cls task is not that critical.

But the detection task does not looks that good.
Even if you take Torch vs OV task at OTE. It look very different, but it should not.

Please see my comparison below for the dog class (raw saliency map, before `get_actmap):`
image
It is exactly the same (I checked pixel-wise).

Could you please take a look into it? There is some bug probably, either in the code or in the visualization.

**Comment by dongkwan-kim01** [comment]
The problem was that I called the `task` several times, and the `background` class in SSD was appended many times so that the cls head shape keeps changing(n_class+1, n_class+2, ...). Totally my bad..

Revised Output:
image
They are identical for pixel-wise(OTX torch == OTE torch == OTE OV).

However OTX OV is still different, maybe because of quantization during compression.

**Review by negvet — APPROVED** [review]

**Review by harimkang — APPROVED** [review]

**Review by goodsong81 — APPROVED** [review]
